File: Java/020_Valid_Parentheses.java
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> map = new HashMap<>();
        Stack<Character> stack = new Stack<>();
        map.put(')', '(');
        map.put(']', '['); 
        map.put('}', '{');
        for (int i = 0; i < s.length(); i++){
            if ("({[".indexOf(s.charAt(i)) != -1){
                stack.push(s.charAt(i));
            } else {
                if (stack.isEmpty() || map.get(s.charAt(i)) != stack.peek()) {
                    return false;
                } else {
                    stack.pop();
                }
            }
        } 
        return stack.isEmpty();     
    }
}

File: Python/020_Valid_Parentheses.py
"""
Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Every close bracket has a corresponding open bracket of the same type.

"""
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        dic = {')':'(', '}':'{', ']':'['}
        for i in s:
            if i in '({[':
                stack.append(i)
            else:
                if len(stack) == 0 or dic[i] != stack[-1]:
                    return False
                else:
                    stack.pop()     
        return len(stack) == 0           